### PR TITLE
drivers: can: flexcan: Fix incorrect timing. Fix for #31014

### LINKED ISSUE #31014: Incorrect timing calculation in can_mcux_flexcan
**Describe the bug**

Fields `propSeg`, `phaseSeg1` and `phaseSeg2` are given to FLEXCAN HAL without conversion (i.e. like register value), but this fields are used in timing calculation like interval value. Interval value is equal register value increased by one.
Incorrect bus speed and large number of errors are in result.

**To Reproduce**
Speed of FlexCAN 250k and greater (When speed is 125k, deviation is not critical).

**Expected behavior**
Real bus speed is equal expected bus speed.

**Impact**
Showstopper

**Environment:**
 - Commit: 4ca7942411f70f0caee6cd872ab14200f88b221b
 - i.MX RT1052

**Variant of solution**
Fields `propSeg`, `phaseSeg1`, `phaseSeg2`, `rJumpwidth` and `preDivider` should decreased by one before giveing to FLEXCAN HAL in lines 140..144 and 171..174 of file `drivers\can\can_mcux_flexcan.c`. Also temporary struct is needed in function `mcux_flexcan_set_timing()`, because `data->timing` should have interval value rather than register value.

### PR DESCRIPTION
**Pull request by alexanderwachter**
The timing values need to be subtracted by one.
Fixes #31014

### CODE REVIEW

**Comment by alexanderwachter** [comment]
ping @GrixaYrev

**Review by mmahadevan108 — APPROVED** [review]

**Comment by MaureenHelm** [comment]
@henrikbrixandersen will you please have a look?

File: drivers/can/can_mcux_flexcan.c [hunk 1]
```diff
@@ -164,14 +167,14 @@ static int mcux_flexcan_set_mode(const struct device *dev, enum can_mode mode)
...
-	       data->timing.phaseSeg2) / data->timing.preDivider;
+	      (1U + data->timing.prop_seg + data->timing.phase_seg1 +
+	       data->timing.phase_seg2) / data->timing.prescaler;
 	flexcan_config.enableIndividMask = true;
 
-	flexcan_config.timingConfig.rJumpwidth = data->timing.rJumpwidth;
-	flexcan_config.timingConfig.propSeg = data->timing.propSeg;
-	flexcan_config.timingConfig.phaseSeg1 = data->timing.phaseSeg1;
-	flexcan_config.timingConfig.phaseSeg2 = data->timing.phaseSeg2;
+	flexcan_config.timingConfig.rJumpwidth = data->timing.sjw - 1U;
```
Comment (karstenkoenig) [inline drivers/can/can_mcux_flexcan.c:174]: This wasn't here before either, but does this code segment not need to set `flexcan_config.timingConfig.preDivider` as well?

Comment (GrixaYrev) [reply drivers/can/can_mcux_flexcan.c:174]: `FLEXCAN_Init()` call `FLEXCAN_SetBaudRate()`, wich calculate `timingConfig.preDivider` (`modules\hal\nxp\mcux\drivers\imx\fsl_flexcan.c:683`).

Comment (karstenkoenig) [reply drivers/can/can_mcux_flexcan.c:174]: Aight, thanks for clearing it up

**Comment by henrikbrixandersen** [comment]
When using a bitrate of 125000 set in DTS, I get the following warning: "[00:00:00.000,000] <wrn> can_mcux_flexcan: Bitrate error: 200". What is a user expected to do?

**Comment by GrixaYrev** [comment]
> When using a bitrate of 125000 set in DTS, I get the following warning: "[00:00:00.000,000] can_mcux_flexcan: Bitrate error: 200". What is a user expected to do?

User should set correct values of `prop-seg`, `phase-seg1` and `phase-seg2` in DTS.

**Comment by alexanderwachter** [comment]
Or just let the algorithm select the timing based on the sample point. The message is telling you that there is in bitrate error of 200 with the given timing.

**Comment by henrikbrixandersen** [comment]
> Or just let the algorithm select the timing based on the sample point. The message is telling you that there is in bitrate error of 200 with the given timing.

This is with the default timing for the twr_ke18 board in Zephyr.

**Comment by alexanderwachter** [comment]
Maybe we could add a shell command that prints timing values. A user can then write them to the device tree to not calculate them online

**Comment by alexanderwachter** [comment]
@henrikbrixandersen we should switch to using the algorithm in device tree. Some of the actual timing in the device tree are just wrong, because they are copied along different boards. They may fit, but sometimes the clock rate changes and then the timing gets wrong (that happened  on as stm32 board already)

**Review by henrikbrixandersen — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**